Add Gene Tracking

lverns opened the pull request:
This pull request brings in the code that allows one to track the individual transfer of genes. It can be enabled and disabled via the `:track-instruction-maps` key in the argmap. When it is enabled, each instruction map in each individual contains a `:uuid`.  Most genes also have a `:parent-uuid` that refers the the instruction map in previous generation that it was copied or mutated from. 

The standard reports in `reports.clj` have been modified to hide this extra information. It is recommended to use EDN printing when this is enabled and while the CSV code runs, I haven't made sure that it still produces valid CSV.

This also makes a small change to the way `:genetic-operators` is added (see commit 44382cd).

Commit:
Add random-insertion flag to random-plush-instruction-map

Commit:
Add random-insertion flag to random-plush-genome-with-size

This brings the functions in random.clj back into a working state, but
there are other calls to random-plush-instruction-map that need to be
updated. Calls to random-plush-genome-with-size do not need to be
changed unless you want the flag to be set.

Commit:
Update call to random-plush-instruction-map in genome_gene_randomize

This commit is not tested.

Commit:
Add random-insertion flag to random-plush-genome

This does not break anything in random.clj, but it most assuredly breaks
code elsewhere. WARNING: The breaking changes are probably not compiler
errors and may show up in strange ways.

Commit:
Update call to random-plush-genome in revert-too-big-child

Commit:
Update call to random-plush-genome in autoconstruction

This commit has not been tested, but it is essentially the same change
as the previous commit so it should work fine.

Commit:
Refactor random-plush-instruction-map

Commit:
Add UUIDs to instruction maps

Commit:
Add function for updating UUIDs on instruction maps

I've only tested this function via the REPL. It is intended to be called
on every individual after it is produced via an operator.

Commit:
Add conditional-thread function

This macro is intended to make the code cleaner when you want to
conditionally thread a value through several functions.

lverns committed:
Remove old commented-out code

lverns committed:
Refactor conditional-thread

Using the or macro is so much cleaner than the anonymous function that I
wrote.

lverns committed:
Fix incorrect comments on conditional-thread

lverns committed:
Add :track-instruction-maps option to argmap

lverns committed:
Make random-plush-instruction-map obey new flag

lverns committed:
Fix bug in perform-genetic-operators

Because the assoc was only being performed when the child wasn't
oversized, and because revert-too-big-child doesn't always set the
:genetic-operator, it was previously possible to produce children with
no :genetic-operator.

lverns committed:
Call update-instruction-map-uuids in perform-genetic-operators

lverns committed:
Pull all the genome printing to a single function

lverns committed:
Hide :uuid and :parent-uuid in standard reports

The information is not readily useable in such a report, the UUIDs make
the logs less readable, and they make the logs take more space in
permanent storage.

Commit:
Merge branch 'feature/detailed-gene-tracing-rebased' into conflict-resolution/detailed-gene-tracing-rebased

The change by @lspector from representing genomes as vectors instead of
lists modified many of the same spots in the code that the gene tracing
did. This merge resolves all the conflicts in a way that preserves the
intent of both changes.

Commit:
Change update-instruction-map-uuids to use vectors

This fixes errors that arise later when the autoconstruction code is
interpreted and tries to manipulate the genomes as vectors which fails
because *this* method turned all the vectors into lazy sequences.

Commit:
Merge branch 'master' into conflict-resolution/detailed-gene-tracing-rebased

This merge had one conflict. Lee moved the argmap into a separate file
and I had added something to the argmap so I had to do the move manually.

NicMcPhee commented on src/clojush/args.clj:
```diff
@@ -177,6 +177,10 @@
           ;; If :silent is used as an epigenetic-marker, this is the probability of random 
           ;; instructions having :silent be true.
+
+          :track-instruction-maps false
```
This seems a non-obvious name to me. I don't necessarily object to it, it just wasn't obvious to me that it meant what the comment said it meant. Thoughts?

lverns replied:
I went with the name because the "things" that we a tracking all come from the `random-plush-instruction-map` function in `src/clojush/random.clj` I guess that calling it `:track-plush-instruction-maps` would be a better mirroring of that. 

We could call it something like `:track-genes`, but that would mean we are defining a "gene" as "an instruction, possibly accompanied by additional information", and I'm not sure if that is the definition that we have been using or not.

NicMcPhee commented on src/clojush/pushgp/breed.clj:
```diff
@@ -82,23 +82,39 @@
+          (fn [genome]
+            (mapv (fn [instruction-map]
+                   (if (:random-insertion instruction-map)
+                     (dissoc instruction-map :random-insertion)
```
Why do we remove that entry? Do we just use the _absence_ of `:parent-uuid` to indicate that it's been randomly generated?

Any how does a randomly inserted gene get it's UUID? Does that happen somewhere else? I'm pretty sure it does, but it's not obvious here so I wanted to ask about it.

lverns replied:
> Why do we remove that entry? Do we just use the absence of :parent-uuid to indicate that it's been randomly generated?

There are two sorts of "randomly generated" genes. The first comes from generation 0, when all are random. None of them get the `:random-insertion` marker. The second sort are randomly inserted by operators. We need to distinguish between the genes that were random in the previous generation and ones are random in this generation. 

Thus we make the operators add the `:random-insertion` flag to all the genes that they create and skip updating the UUIDs on them. However, we need to strip of the flag so that in the next generation they will get properly updated.

Does that help at all?

> An[d] how does a randomly inserted gene get it's UUID?

All genes are ultimately created by `random-plush-instruction-map` which gives them a UUID.

NicMcPhee commented on src/clojush/pushgp/genetic_operators.clj:
```diff
@@ -350,7 +350,7 @@ be set globally or eliminated in the future."
         new-genome (if variant
                      child-genome
-                     (random-plush-genome max-genome-size-in-initial-program atom-generators argmap))]
+                     (random-plush-genome max-genome-size-in-initial-program atom-generators true argmap))]
```
It seems awkward that we introduce an unnamed `true` here. Everything else here has a name that at least sort of suggests what it's doing, but that `true` just floats there without any connection to the world in a way that makes it hard to understand. Is there a (reasonable) way we could have this have a name here?

NicMcPhee commented on src/clojush/pushgp/report.clj:
```diff
@@ -41,6 +41,9 @@
       (println (name param) "=" val))))
 
+(defn print-genome [individual]
+  (pr-str (not-lazy (map #(dissoc % :uuid :parent-uuid) (:genome individual)))))
```
I realize that you inherited the `not-lazy` here, but I can't quite figure out why we need. I assume this is here because of @lspector's paranoia about laziness, but I don't see how laziness is an issue here. Since the string this generates in printed (immediately), that will force the list to be evaluated, right? Or is there something with threading that can act weirdly here?

Also, (and this isn't really an issue for this PR), this suggests that we might want some more configurability in specifying which fields you'd want to print in the reporting.

lspector replied:
I'm not sure if/how it would really manifest here, but this is about something different from my normal paranoia about laziness (which has to do with whether GC will properly reclaim storage for unrealized lazy sequences). The issue here relates specifically to printing. Sometimes, when you do some printing things, or converting to string things, with lazy sequences, you get a weird internal representation like "clojure.lang.LazySeq@c5d38b66", even if the sequence is fully evaluated. I do not fully understand what's going on here. It may be that using `pr-str` takes care of this, but I am not sure. To be safe, however, and because it is so deeply annoying to ever see something like "clojure.lang.LazySeq@c5d38b66" when you really wanted to see what's in the sequence, I have often taken to sprinkling `not-lazy` liberally through any code that prints strings from sequences.

NicMcPhee commented on src/clojush/random.clj:
```diff
@@ -38,50 +38,84 @@
         (recur (inc parens)
                (rest probabilities))))))
 
+(defn conditional-thread
```
Did you invent this, or does this come from somewhere? 

Am I correct in thinking you could use `cond->` for this, but you'd end up with a bunch of repetition since you'd have to repeat each expression as both the "test" and the "result"?

Does it belong here, or in the `util`?

lverns replied:
> Am I correct in thinking you could use `cond->` for this, but you'd end up with a bunch of repetition since you'd have to repeat each expression as both the "test" and the "result"?

In the two places where `conditional-thread` is actually used, `cond->` would work fine, so I'll make the switch and remove this.

> Does it belong here, or in the `util`?

Probably in `util`, but I'd have to double check namespaces.

NicMcPhee commented on src/clojush/random.clj:
```diff
@@ -38,50 +38,84 @@
-                   markers)))))
+   (random-plush-instruction-map atom-generators {}))
+  ([atom-generators argmap]
+   (random-plush-instruction-map atom-generators false argmap))
```
Again, the floating constant `false` is hard to interpret here.

Also a lot of this diff appears to be white space. Can we reformat so it matches the original code so it's easier to see what's really different.

lverns replied:
> Also a lot of this diff appears to be white space. Can we reformat so it matches the original code so it's easier to see what's really different.

Huh, I think you're right. Looking at this diff, it isn't clear why indentation should have changed. I'll see if I can correct that.

NicMcPhee commented on src/clojush/random.clj:
```diff
@@ -38,50 +38,84 @@
-                      atom-generators
-                      argmap))))
+  ([genome-size atom-generators argmap]
+   (random-plush-genome-with-size genome-size atom-generators false argmap))
```
`false` again.

NicMcPhee commented on src/clojush/random.clj:
```diff
@@ -38,50 +38,84 @@
   ([max-genome-size atom-generators]
     (random-plush-genome max-genome-size atom-generators {}))
   ([max-genome-size atom-generators argmap]
+   (random-plush-genome max-genome-size atom-generators false argmap))
```
`false` one last time. 😄

NicMcPhee reviewed (changes requested):
This mostly looks really nice. Most of my notes are more random questions than serious concerns. There are a few things, though, that we might want to address (like the location of `conditional-thread`).

lverns reviewed (commented):
I've added responses most of your comments and will start working on the changes. As per our conversation on Tuesday, we'll want to take a little time looking at the code together to see if there is anything that could reduce/make more clear the floating boolean values.

Commit:
Replace conditional-thread with cond->

Using forms built in to Clojure is preferred over custom code.

Commit:
Undo whitespace changes

This will not make `git blame` more useful, but it will make the diff
more readable when merging into master. The best thing to do would be to
go back and modify commit 015d591836425f3ba6e3256a0e8480c3ae399f07 to
not introduce the whitespace, but I'm not sure that we can do that at
this point.

Commit:
Merge branch 'master' into conflict-resolution/detailed-gene-tracing-rebased

This brings in a whole bunch of new code (mainly operators) that will
need to be adapted to work with the gene tracking before this can be
merged to master.

Commit:
Remove remove-insertion flag from random-plush-instruction-map

This changes the code introduced in
015d591836425f3ba6e3256a0e8480c3ae399f0. I didn't do a `git revert`
because it needed to be done manually.

Commit:
Remove random-instruction flag from random-plush-genome-with-size

Also, adjust formatting of this function so that it matches that on
master.

Commit:
Remove random-insertion flag from random-plush-genome

Commit:
Update calls to random-plush-genome

They now use a consistent number of arguments (and the number is aligns
with what random-plush-genome expects).

Commit:
Update calls to random-plush-instruction-map

Commit:
Merge branch 'master' into conflict-resolution/detailed-gene-tracing-rebased

lverns commented:
This pull request was just created to make the code reviewing easier to do and that's done now.
